Hotfix wrong override when value is not defined in breakpoint

AntoineDuComptoirDesPharmacies opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?

This PR tend to avoid generating a wrong CSS override when value is not known in theme breakpoint (Same as for `gap`, etc...)

#### Where should the reviewer start?

Read the linked issue to see the difference between both Codepens.

#### What testing has been done on this PR?

Manual tests only

#### How should this be manually tested?

Use Developper console with the new version to figure out that CSS override have not been generated for unknown value.
Then add a 'small' breakpoint with distinct edgeSize value to see the override being generated.

#### Do Jest tests follow these best practices?

No Jest tests.

#### Any background context you want to provide?

#### What are the relevant issues?

Issue #6968 

#### Screenshots (if appropriate)

N/A

#### Do the grommet docs need to be updated?

No

#### Should this PR be mentioned in the release notes?

Maybe, depending if a lot of people get a specific expected behavior with this bug.

#### Is this change backwards compatible or is it a breaking change?

It should be backwards compatible, except if some users get a specific behavior getting the raw CSS value inserted and being different than the value defined in global theme.

AntoineDuComptoirDesPharmacies committed:
Hotfix wrong override when value is not defined in breakpoint

AntoineDuComptoirDesPharmacies committed:
Run yarn jest -u as requested by @britt6612.

AntoineDuComptoirDesPharmacies commented:
> @AntoineDuComptoirDesPharmacies can you run `yarn jest -u` to update the snapshot?

Done, thank you for this tip !

britt6612 commented:
looking into the jest test now and seeing what going on.. for the bundle size error can you go [here](https://github.com/grommet/grommet/blob/master/package.json#L208) and change `bundle size` to 162 and that at least fixes that one error I while I look into the jest error!

britt6612 commented:
@AntoineDuComptoirDesPharmacies I beleive your `yarn.lock` is not in sync with master if you can `git pull upstream master` then run `yarn` and then ` yarn jest -u` it should work. Let me know if you have any issues

AntoineDuComptoirDesPharmacies committed:
Update yarn lock and update yarn + max package size

AntoineDuComptoirDesPharmacies committed:
Merge remote-tracking branch 'origin/master' into fix-breakpoint-default-edge-size

AntoineDuComptoirDesPharmacies committed:
Fix tests

jcfilben commented:
You can also try creating a new fresh branch to see if that resolves the snapshot issues

AntoineDuComptoirDesPharmacies committed:
Merge remote-tracking branch 'origin/master' into fix-breakpoint-default-edge-size

AntoineDuComptoirDesPharmacies committed:
Revert "Fix tests"This reverts commit a4022084e1adcb8aa58a9f5657f25f679a854a1c.Revert "Update yarn lock and update yarn + max package size"This reverts commit 2c249a210207a3fb31de4681d5daf353e15bcfc1.

Revert "Run yarn jest -u as requested by @britt6612."

This reverts commit 8ac167c6609f745d1022c11b8e0de7e4435d82a9.

AntoineDuComptoirDesPharmacies committed:
Put again 162 kB as it is modified in master

AntoineDuComptoirDesPharmacies committed:
Remove useless style in tests

AntoineDuComptoirDesPharmacies commented:
> You can also try creating a new fresh branch to see if that resolves the snapshot issues

Thank you, I revert the commits containing snapshot update and modify the only snapshot who have changed.
We are all good now concerning tests.

taysea commented on src/js/utils/styles.js:
```diff
@@ -41,17 +41,19 @@ export const edgeStyle = (
-  const breakpoint =
-    responsiveBreakpoint && theme.global.breakpoints[responsiveBreakpoint];
+  const breakpoint = getBreakpointStyle(theme, responsiveBreakpoint);
+  const metric = theme.global.edgeSize[data] || data;
```
Behavior looks great. Two small naming suggestions would be to align the variable names with the concept of CSS value. Metric was a little bit confusing at first:
- metric --> value
- responsiveMetric --> responsiveValue

soft opinion.

AntoineDuComptoirDesPharmacies replied:
Hi @taysea,
I agree with you. However 'value' will conflict with the variable 'value' used for horizontalVerticalEqual (see line 77).
I propose to rename 'value' to 'commonValue'. Just pushed the modification !

taysea commented on src/js/components/Tip/__tests__/__snapshots__/Tip-test.tsx.snap:
```diff
@@ -579,12 +579,6 @@ exports[`Tip themed 1`] = `
   animation-delay: 0.01s;
 }
 
-@media only screen and (max-width:768px) {
```
I'm trying to understand why this margin goes away with these changes. Can you explain?

AntoineDuComptoirDesPharmacies replied:
Previously, this responsive style was generated by the code because the value for '**margin**' is not present in the responsive breakpoint definition of '**_edgeSize_**', resulting in raw value inserted in the CSS.

However in this context, this custom responsive style is clearly useless as this margin value should be applied to every viewport width. We can see that margin value is already in the common style here : https://github.com/grommet/grommet/pull/6969/files#diff-78a408ca8784746e3a90eddb67f1909053f9555be32115f0d0e68f5b5fa759e0R522

This custom responsive section should only appear if, depending theme breakpoints, the value is changing according to viewport width.

For example, if we are replacing '21px' with 'small', the code will generate a custom section for  max-width:768px with a value of '6px' where the common CSS will have a value of '12px' : 
```
      <Grommet
        theme={{
          tip: {
            drop: {
              background: 'brand',
              elevation: 'large',
              margin: 'small',
            },
          },
        }}
      >
        <Tip content="tooltip">Example</Tip>
      </Grommet>
```

![image](https://github.com/user-attachments/assets/3431ea9f-573f-4d19-a0a5-387df661f622)

I hope my explanation is clear enough. Do not hesitate to tell me if not.

britt6612 replied:
@AntoineDuComptoirDesPharmacies does this fix only cover the typeof data === 'string' path? The object form branches `horizontal` `vertical` `top` `bottom` still use `breakpoint.edgeSize[x] || x]` which would fall back to the raw token name if the breakpoint doesn't define that size. Should we extend the same guard to those branches as well?

AntoineDuComptoirDesPharmacies replied:
Hi @britt6612,
You are right ! This was raise too by the Copilot reviewer.
I did a commit ([6969](https://github.com/grommet/grommet/pull/6969/changes)) to fix the problem for branches `horizontal`, `vertical`, `top`, `bottom`, `allEdges`.

I would like to send you a Codesandbox with the example fixed but Codesandbox is down since one week...

Yours faithfully,
LCDP

AntoineDuComptoirDesPharmacies replied:
Here is the codesandbox with problem reproduced : 
https://codesandbox.io/p/devbox/grommet-v2-template-forked-nlw72p?workspaceId=ws_3XaV72QGZZgxUDQujYFsKj

Here is a codesandbox with the last commit fixing these behaviors : 
https://codesandbox.io/p/devbox/grommet-v2-template-forked-fjmwsl?workspaceId=ws_3XaV72QGZZgxUDQujYFsKj

britt6612 replied:
@AntoineDuComptoirDesPharmacies thank you very much for the codesandbox this helps thanks for addressing!

AntoineDuComptoirDesPharmacies committed:
Rename metric and responsiveMetric according to PR remarks from @taysea

AntoineDuComptoirDesPharmacies commented:
Hi @taysea,

Thank you for this review.
I just modified the code and answer to your comments.

Should I remerge 'master' branch into mine to solve potential conflicts ?

Yours faithfully,
LCDP

AntoineDuComptoirDesPharmacies commented:
Hi @taysea and @jcfilben,

Do you know what is the next step to continue into the approval process please ?

Thanks in advance for your help !
Yours faithfully,
LCDP

britt6612 commented:
> Hi @taysea and @jcfilben,
> 
> Do you know what is the next step to continue into the approval process please ?
> 
> Thanks in advance for your help ! Yours faithfully, LCDP

@AntoineDuComptoirDesPharmacies Sorry this fell of our radar! It looks like there might be some conflicts within this branch (that may be why things are failing) can you look into that then we can put this back on our list to review!

AntoineDuComptoirDesPharmacies committed:
Merge branch 'master' into fix-breakpoint-default-edge-size

# Conflicts:
#	src/js/components/Tip/__tests__/__snapshots__/Tip-test.tsx.snap

AntoineDuComptoirDesPharmacies commented:
Hi @britt6612,

No problem, I just resolved the PR against last commit of master.

Yours faithfully,
LCDP

AntoineDuComptoirDesPharmacies committed:
Merge branch 'master' into fix-breakpoint-default-edge-size

AntoineDuComptoirDesPharmacies committed:
#6969

Take into account Copilot remarks
Apply the fix on complex object in addition to String value

Signed-off-by: Antoine <antoine@lecomptoirdespharmacies.fr>

AntoineDuComptoirDesPharmacies committed:
Merge remote-tracking branch 'origin/fix-breakpoint-default-edge-size' into fix-breakpoint-default-edge-size

AntoineDuComptoirDesPharmacies commented:
Hi @jcfilben, for information I took into account Copilot remarks.

Yours faithfully,

jcfilben commented:
@AntoineDuComptoirDesPharmacies looks like @britt6612 left a question here: https://github.com/grommet/grommet/pull/6969#discussion_r3119846564

AntoineDuComptoirDesPharmacies commented:
> @AntoineDuComptoirDesPharmacies looks like @britt6612 left a question here: [#6969 (comment)](https://github.com/grommet/grommet/pull/6969#discussion_r3119846564)

Hi @jcfilben, comment processed ! 👍

jcfilben reviewed (approved):
Looks good!

jcfilben commented:
@AntoineDuComptoirDesPharmacies could you update your branch with the latest version of the master branch? Then I can merge

AntoineDuComptoirDesPharmacies committed:
Merge branch 'grommet:master' into fix-breakpoint-default-edge-size
